lib/classes/plugin-manager.js

```diff
@@ -83,13 +81,13 @@ class PluginManager {
     this.cliCommands = commands;
   }
 
-  addPlugin(Plugin) {
+  addPlugin(Plugin, options = {}) {
     const pluginUtils = {};
-    if (Plugin._serverlessExternalPluginName) {
-      Object.assign(pluginUtils, getPluginWriters(Plugin._serverlessExternalPluginName));
+    if (!options.isInternal) {
+      Object.assign(pluginUtils, getPluginWriters(options.name || Plugin.name));
```

Comment:
Indeed I forgot that `_serverlessExternalPluginName` hosts name we need.

Nonetheless let's follow the route I proposed, but keep track of current external plugin name (those which don't have `_serverlessExternalPluginName` are plugins which are initialized internally by last one that has  `_serverlessExternalPluginName`, so we should get right result) .

See following:

```javascript
let currentExternalPluginName;
...
addPlugin(Plugin) {
  ...
  if (isRegisteringExternalPlugins) {
    if (Plugin._serverlessExternalPluginName) currentExternalPluginName = Plugin._serverlessExternalPluginName;
    Object.assign(pluginUtils, getPluginWriters(currentExternalPluginName));
  }
```

Reply:
How does `currentExternalPluginName` get set?

Reply:
See example I pasted in comment above, it's set there

Reply:
Consider this `serverless.yml`

``` yaml
service: foo
plugins:
    - my-plugin
```

This is `my-plugin`:

``` js
const WarmupPlugin = require("serverless-plugin-warmup");

class MyPlugin {
    constructor(serverless) {
        this.serverless = serverless;
        this.hooks = {
            initialize: () => this.init(),
        }
    }

    init() {
        this.serverless.pluginManager.addPlugin(WarmupPlugin);
    }
}

module.exports = MyPlugin;
```

This is what will happen:

1. serverless assigns `MyPlugin._serverlessExternalPluginName = 'my-plugin'`;
2. serverless calls `addPlugin(MyPlugin)`.
3. `currentExternalPluginName` gets set to to `my-plugin`.
4. `MyPlugin#init()` gets called and executes `addPlugin(WarmupPlugin)`.
5. Since `WarmupPlugin` doesn't have `_serverlessExternalPluginName` it will use `currentExternalPluginName`.
6. `WarmupPlugin` logger uses `my-plugin` as the `pluginName` instead of `serverless-plugin-warmup`.


My PR lets my `MyPlugin#init()` to pass the plugin name to `addPlugin`:

``` js
const WarmupPlugin = require("serverless-plugin-warmup");

class MyPlugin {
    constructor(serverless) {
        this.serverless = serverless;
        this.hooks = {
            initialize: () => this.init(),
        }
    }

    init() {
        this.serverless.pluginManager.addPlugin(WarmupPlugin, { name: 'serverless-plugin-warmup' });
    }
}
```

Reply:
> MyPlugin#init() gets called and executes addPlugin(WarmupPlugin).
> Since WarmupPlugin doesn't have _serverlessExternalPluginName it will use currentExternalPluginName

That's expected.

If you want to some different name on `WarmupPlugin`, simply hack it to:  `WarmupPlugin._serverlessExternalPluginName`.

Note that `pluginManger.addPlugin` is considered a _private_ not _public_ part of the Framework (it's even not officially documented), and this is a very core part of the Framework, we do not want to introduce any big changes here

Reply:
What if we just used the plugin constructor name as the fallback?

``` js
if (isRegisteringExternalPlugins) {
    const name = Plugin._serverlessExternalPluginName || Plugin.name;
    Object.assign(pluginUtils, getPluginWriters(name));
}
```

Reply:
I think it'll be better to stick to `Plugin._serverlessExternalPluginName`, as in all other known to me cases, it's preferable that we pass the same logger as to the top plugin.

Reply:
Can we at least do something like this so it doesn't explode?

``` js
if (isRegisteringExternalPlugins) {
    Object.assign(pluginUtils, getPluginWriters(Plugin._serverlessExternalPluginName || 'unknown'));
}
```

Reply:
> Can we at least do something like this so it doesn't explode?

I think it wouldn't as `_serverlessExternalPluginName` is set upfront on processing the first top-level plugin and never cleared

Reply:
Alright, I guess I'll just stick with this hack then:

``` js
const WarmupPlugin = require("serverless-plugin-warmup");

class MyPlugin {
    constructor(serverless) {
        this.serverless = serverless;
        this.hooks = {
            initialize: () => this.init(),
        }
    }

    init() {
        WarmupPlugin._serverlessExternalPluginName = 'serverless-plugin-warmup';
        this.serverless.pluginManager.addPlugin(WarmupPlugin);
    }
}
```

Reply:
> Note that pluginManger.addPlugin is considered a private not public part of the Framework (it's even not officially documented)

FWIW, there's hardly any plugin dev documentation.